client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -361,11 +361,19 @@ public void cancelled() {
     private void setHeaders(HttpRequest httpRequest, Header[] requestHeaders) {
         Objects.requireNonNull(requestHeaders, "request headers must not be null");
         for (Header defaultHeader : defaultHeaders) {
-            httpRequest.setHeader(defaultHeader);
+            httpRequest.addHeader(defaultHeader);
         }
-        for (Header requestHeader : requestHeaders) {
-            Objects.requireNonNull(requestHeader, "request header must not be null");
-            httpRequest.setHeader(requestHeader);
+        if (requestHeaders.length != 0) {
+            final Set<String> uniqueNames = new HashSet<>();
+            for (Header requestHeader : requestHeaders) {
+                Objects.requireNonNull(requestHeader, "request header must not be null");
+                // overwrite any default header
+                if (uniqueNames.add(requestHeader.getName())) {
+                    httpRequest.setHeader(requestHeader);
+                } else {
+                    httpRequest.addHeader(requestHeader);
```

Comment:
If a user sets the default headers as something like:

```
Authorization: Basic blahBlahBlah
```

This would represent some generic user (e.g, readonly access). However, for certain requests, they may need to elevate their level of access, so they provide

```
Authorization: Basic blahBlahBlah2
```

as a request-level header.

If we use `add` all the way, then the raw request will have:

```
Authorization: Basic blahBlahBlah
Authorization: Basic blahBlahBlah2
```

By using `set` the first time we come across a request-level header, it throws away any header set by the default.

Perhaps I could rewrite this to change the order, thereby always using `add`?

1. write all Request Headers
2. write any Default Headers not already written